### PR TITLE
Uniffi-rs migration

### PR DESCRIPTION
**Pull request by juanky201271**
This is the first step to use uniffi-rs, and of course there are a lot space to grow and improve.

- libuniffi_rustlib working in Android - Kotlin bindings
- libuniffi_rustlib working in IOS - Swift bindings.
- IOS Background Syncing in Swift.
- Android & IOS Test version 1.3.8 (163).

Ref:
- https://reactnative.dev/docs/native-modules-ios#exporting-swift
- https://mozilla.github.io/uniffi-rs/

### CODE REVIEW

File: android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt [hunk 1]
```diff
@@ -110,18 +110,18 @@ class ExecuteAddressesFromSeed {
         val server = "http://10.0.2.2:20000"
         val chainhint = "main"
         val seed = Seeds.ABANDON
-        val birthday = "1"
+        val birthday:ULong = 1u
```
Comment (zancas) [inline android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt:113]: Nice!

File: android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt [hunk 2]
```diff
@@ -110,18 +110,18 @@ class ExecuteAddressesFromSeed {
...
         val datadir = MainApplication.getAppContext()!!.filesDir.path
-        val monitorMempool = "false"
+        val monitorMempool = false
```
Comment (zancas) [inline android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt:115]: Nice.

File: rust/Cargo.toml [hunk 3]
```diff
@@ -8,9 +8,10 @@ members = [
 resolver = "2"
 
 [workspace.dependencies]
-log = "0.4.8"
 zingolib = { git="https://github.com/zingolabs/zingolib", default-features=true, branch = "stable" }
 zingoconfig = { git="https://github.com/zingolabs/zingolib", default-features=true, branch = "stable" }
+uniffi = { version = "0.27.0" }
```
Comment (zancas) [inline rust/Cargo.toml:12]: In general let's allow the minor `X.Y.Z` the ("Z") of semver to auto-update by specifying deps like this:

`dep = { version = "X.Y" }`

This will mean that cargo update is free to select the latest minor version, so we'll automatically get security updates.

Comment (zancas) [reply rust/Cargo.toml:12]: @AloeareV does using only the first two elements of the semver string make sense to you?

File: rust/android/Cargo.toml [hunk 4]
```diff
@@ -8,22 +8,12 @@ edition = "2021"
...
-rustlib = { path = "../lib" }
-
 [dev-dependencies]
-tokio =  { version = "1.24.2", features = ["full"] }
 zingoconfig = { workspace = true }
+zingomobile_utils = { path = "../zingomobile_utils" }
+regchest_utils = { git="https://github.com/zingolabs/zingo-regchest", default-features=true, branch = "dev" }
 zingo-testutils = { git="https://github.com/zingolabs/zingolib", default-features=true, branch = "stable" }
 darkside-tests = { git="https://github.com/zingolabs/zingolib", default-features=true, branch = "stable" }
+tokio =  { version = "1.24.2", features = ["full"] }
```
Comment (zancas) [inline rust/android/Cargo.toml:17]: I propose that we use: `tokio = { version = "1.24", features = ["full"] }`

What do you think @juanky201271 @AloeareV ?

File: rust/ios/Cargo.toml [hunk 5]
```diff
@@ -5,8 +5,7 @@ authors = ["Zingolabs <zingo@zingolabs.com>"]
 edition = "2021"
 
 [dependencies]
-log = { workspace = true }
-rustlib = { path = "../lib" }
+rustlib = { workspace = true }
```
Comment (zancas) [inline rust/ios/Cargo.toml]: Nice!

File: ios/RPCModuleSwift.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
+//
+//  RPCModule.swift
+//  ZingoMobile
+//
+//  Created by Juan Carlos Carmona Calvo on 4/2/24.
```
Comment (zancas) [inline ios/RPCModuleSwift.swift]: You're rocking!

File: rust/ios/src/lib.rs [hunk 7]
```diff
@@ -1,378 +0,0 @@
-use std::ffi::{CStr, CString};
-use std::os::raw::c_char;
-
-#[no_mangle]
-pub extern "C" fn rust_free(s: *mut c_char) {
```
Comment (juanky201271) [inline rust/ios/src/lib.rs:5]: @AloeareV do you think this method make any sense to copy/create down in the unique rust layer now? It seems like for the removed rust ios layer, it clean the memory...

File: app/LoadedApp/LoadedApp.tsx [hunk 8]
```diff
@@ -346,28 +346,41 @@ export class LoadedAppClass extends Component<LoadedAppClassProps, AppStateLoade
...
+      console.log('LOADED', 'prior', this.state.appState, 'next', nextAppState);
       if (Platform.OS === 'ios') {
-        if (this.state.appState === 'background' && nextAppState === 'inactive') {
+        if ((this.state.appState === 'inactive' && nextAppState === 'active') || (this.state.appState === 'active' && nextAppState === 'inactive')) {
+          console.log('LOADED SAVED IOS do nothing', nextAppState);
+          this.setState({ appState: nextAppState });
           return;
         }
-        if (this.state.appState === 'inactive' && nextAppState === 'active') {
+        if (this.state.appState === 'inactive' && nextAppState === 'background') {
```
Comment (pacu) [inline app/LoadedApp/LoadedApp.tsx:359]: all of these strings should be moved to constants. 
see https://www.typescriptlang.org/docs/handbook/enums.html

File: app/LoadedApp/LoadedApp.tsx
Comment (pacu) [inline app/LoadedApp/LoadedApp.tsx:1]: I know this is outside the scope of the PR but this code could be significantly improved by extracting the nextAppState closure to an object that handles the state and testing that the transitions that are expected actually happen.

File: components/Send/Send.tsx [hunk 9]
```diff
@@ -277,20 +277,10 @@ const Send: React.FunctionComponent<SendProps> = ({
...
+      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
     });
     const keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', () => {
-      Animated.timing(slideAnim, {
-        toValue: 0,
-        duration: 100,
-        easing: EasingNode.linear,
-        //useNativeDriver: true,
-      }).start();
+      slideAnim.value = withTiming(0, { duration: 100, easing: Easing.linear });
```
Comment (pacu) [inline components/Send/Send.tsx:283]: these changes are not uniffi-rs related, are they? If they are not. I'd suggest that you remove them and do them in another PR

File: components/Seed/Seed.tsx [hunk 10]
```diff
@@ -137,20 +137,10 @@ const Seed: React.FunctionComponent<SeedProps> = ({ onClickOK, onClickCancel, ac
 
   useEffect(() => {
     const keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', () => {
-      Animated.timing(slideAnim, {
-        toValue: 0 - titleViewHeight + 25,
-        duration: 100,
-        easing: EasingNode.linear,
-        //useNativeDriver: true,
-      }).start();
+      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
```
Comment (pacu) [inline components/Seed/Seed.tsx:140]: these changes are not uniffi-rs related, are they? If they are not. I'd suggest that you remove them and do them in another PR

File: components/Memo/Memo.tsx [hunk 11]
```diff
@@ -56,20 +56,10 @@ const Memo: React.FunctionComponent<MemoProps> = ({ closeModal, updateToField })
 
   useEffect(() => {
     const keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', () => {
-      Animated.timing(slideAnim, {
-        toValue: 0 - titleViewHeight + 25,
-        duration: 100,
-        easing: EasingNode.linear,
-        //useNativeDriver: true,
-      }).start();
+      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
```
Comment (pacu) [inline components/Memo/Memo.tsx:59]: these changes are not uniffi-rs related, are they? If they are not. I'd suggest that you remove them and do them in another PR

File: components/Settings/Settings.tsx [hunk 12]
```diff
@@ -186,20 +186,10 @@ const Settings: React.FunctionComponent<SettingsProps> = ({
 
   useEffect(() => {
     const keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', () => {
-      Animated.timing(slideAnim, {
-        toValue: 0 - titleViewHeight + 25,
-        duration: 100,
-        easing: EasingNode.linear,
-        //useNativeDriver: true,
-      }).start();
+      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
```
Comment (pacu) [inline components/Settings/Settings.tsx:189]: these changes are not uniffi-rs related, are they? If they are not. I'd suggest that you remove them and do them in another PR

File: components/Ufvk/ImportUfvk.tsx [hunk 13]
```diff
@@ -35,24 +35,14 @@ const ImportUfvk: React.FunctionComponent<ImportUfvkProps> = ({ onClickCancel, o
...
 
   useEffect(() => {
     const keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', () => {
-      Animated.timing(slideAnim, {
-        toValue: 0 - titleViewHeight + 25,
-        duration: 100,
-        easing: EasingNode.linear,
-        //useNativeDriver: true,
-      }).start();
+      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
```
Comment (pacu) [inline components/Ufvk/ImportUfvk.tsx:42]: these changes are not uniffi-rs related, are they? If they are not. I'd suggest that you remove them and do them in another PR

File: ios/AppDelegate.swift [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+        // Iniciar la sincronización
+        NSLog("BGTask startBackgroundTask run sync task")
+        // Para ejecutar solo una tarea
+        //DispatchQueue.global(qos: .background).async {
+        self.syncingProcessBackgroundTask()
+        //}
+
+        task.expirationHandler = {
+            NSLog("BGTask startBackgroundTask - expirationHandler called")
+            // Interrumpir el proceso de sincronización, no puedo esperar a ver si el proceso ha terminado
```
Comment (pacu) [inline ios/AppDelegate.swift]: Love the comments in Spanish. Unfortunately it is not the _Lingua Franca_  of most projects. I'd consider translating these.

File: ios/RPCModule.swift [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+            }
+            return seedStr
+          } else {
+            NSLog("Error creating new wallet")
+            return ""
+          }
+      }
+  }
+
+  @objc(createNewWallet:chainhinter:resolver:rejecter:)
```
Comment (pacu) [inline ios/RPCModule.swift]: I'm curious about the 'er' suffix of the parameters like `chainhinter`

File: ios/RPCModule.swift [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+  @objc(restoreWalletFromUfvk:birthdayer:serverer:chainhinter:resolver:rejecter:)
+  func restoreWalletFromUfvk(restoreUfvk: String, birthday: String, server: String, chainhint: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
+      DispatchQueue.global().async {
+          let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
+          if let documentsDirectory = paths.first {
+              let ufvk = initFromUfvk(serveruri: server, ufvk: restoreUfvk, birthday: UInt64(birthday) ?? 0, datadir: documentsDirectory, chainhint: chainhint, monitorMempool: true)
+              let ufvkStr = String(ufvk)
+              if !ufvkStr.lowercased().hasPrefix("error") {
+                self.saveWalletInternal()
```
Comment (pacu) [inline ios/RPCModule.swift]: this is causing a memory cycle between this closure passed to `DispatchQueue.global().async()` and `self`

you should do this
````Swift
  DispatchQueue.global().async { [weak self]
````

see https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/

File: ios/RPCModule.swift [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+              resolve(ufvkStr)
+          } else {
+              NSLog("Error restoring wallet from ufvk")
+              resolve("")
+          }
+      }
+  }
+
+  func loadExistingWallet(server: String, chainhint: String) -> String {
+      autoreleasepool {
```
Comment (pacu) [inline ios/RPCModule.swift]: I'm not sure whether these calls to `autoreleasepool` are contributing  much

File: ios/RPCModule.swift [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+            }
+          }
+      }
+  }
+
+  @objc(execute:argser:resolver:rejecter:)
+  func execute(method: String, args: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
+      let dict: [String: Any] = ["method": method, "args": args, "resolve": resolve]
+      DispatchQueue.global().async {
+          self.doExecuteOnThread(dict)
```
Comment (pacu) [inline ios/RPCModule.swift]: same thing here. reference to self should be weakened

File: ios/RPCModule.swift [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+              resolve("")
+            }
+          }
+      }
+  }
+
+  @objc(execute:argser:resolver:rejecter:)
+  func execute(method: String, args: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
+      let dict: [String: Any] = ["method": method, "args": args, "resolve": resolve]
+      DispatchQueue.global().async {
```
Comment (pacu) [inline ios/RPCModule.swift]: I'm not sure this is the exact equivalent of `NSThread detachNewThreadSelector` which is used in RPCModule.m

https://stackoverflow.com/questions/5653522/grand-central-dispatch-vs-nsthread
Although in theory GCD should efficiently dispatch the queued tasks and be in average more performant and reliable that using NSThread, I don't know how this change affects ReactNative's assumptions.

File: ios/RPCModuleBridge.m [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+    (NSString)restoreSeed 
+    birthdayer:(NSString)birthday 
+    serverer:(NSString)server 
+    chainhinter:(NSString)chainhint 
+    resolver:(RCTPromiseResolveBlock)resolve 
+    rejecter:(RCTPromiseRejectBlock)reject)
+RCT_EXTERN_METHOD(restoreWalletFromUfvk: 
+    (NSString)restoreUfvk 
+    birthdayer:(NSString)birthday 
+    serverer:(NSString)server
```
Comment (pacu) [inline ios/RPCModuleBridge.m]: these `er` suffixes are weird

File: rust/lib/src/rustlib.udl [hunk 21]
```diff
@@ -0,0 +1,39 @@
+namespace rustlib {
+  string init_logging();
+  string execute_command(string cmd, string args);
```
Comment (pacu) [inline rust/lib/src/rustlib.udl:3]: This is a good start for UniFFI and a good judgement to approach it. 

IDK if the intention is to keep growing this API and start extracting the string commands of `execute_commands` into their own methods or stick with the local client/server approach that is currently being used. 

I think the succinct API that the current approach has is convenient in terms of boilerplate code that is needed to expose changes in the rust library.  The downside is that the app relies on flying strings. like `execute_command(cmd: "sync", "")` and this returns some other string 

A more robust API would having a `rustlib.sync()` that returns a `ReturnType` result instead of a string. The downside is that it envolves coding the intermediate code. Maybe the scaffolding created from the udl file can take care of most of it. 

This can be also remediated by creating a better TypeScript API for interacting with `execute_command` and keep this succinct UniFFI API as is to keep a better velocity and less boiler plate for exposing new functionality on the app.

The main problem is that the Strings resulting on the responses could be broken from a change in the zingolib without that causing a problem in the build itself.

**Review by pacu — CHANGES_REQUESTED** [review]

File: app/LoadedApp/LoadedApp.tsx (hunk 8 above)
Comment (juanky201271) [reply app/LoadedApp/LoadedApp.tsx:359]: Absolutely... I will review all of these `strings` and create constants... but I will do in another PR, because there are more in the code-base.

File: components/Send/Send.tsx (hunk 9 above)
Comment (juanky201271) [reply components/Send/Send.tsx:283]: I needed to upgrade `react-native-reanimated` from 2.x.x to 3.7.0 because some problem with this PR... and those changes are necessary to make it work in the new version.

File: components/Seed/Seed.tsx (hunk 10 above)
Comment (juanky201271) [reply components/Seed/Seed.tsx:140]: same

File: ios/RPCModule.swift (hunk 17 above)
Comment (juanky201271) [reply ios/RPCModule.swift]: in obj-c it seems like helps with memory releasing... I hope in swift works the same way.

File: components/Memo/Memo.tsx (hunk 11 above)
Comment (juanky201271) [reply components/Memo/Memo.tsx:59]: same

File: components/Settings/Settings.tsx (hunk 12 above)
Comment (juanky201271) [reply components/Settings/Settings.tsx:189]: same

File: components/Ufvk/ImportUfvk.tsx (hunk 13 above)
Comment (juanky201271) [reply components/Ufvk/ImportUfvk.tsx:42]: same

File: ios/RPCModule.swift (hunk 15 above)
Comment (juanky201271) [reply ios/RPCModule.swift:201]: It seems like the objc() needs some king of tags to refer the params... like `rejecter`, `resolver`, and following this pattern: `serverer`, `chainhinter`... Maybe I can remove all of these tags, and leave only the params names... What do you think? @pacu

File: rust/lib/src/rustlib.udl (hunk 21 above)
Comment (juanky201271) [reply rust/lib/src/rustlib.udl:3]: > This is a good start for UniFFI and a good judgement to approach it.
> 
> IDK if the intention is to keep growing this API and start extracting the string commands of `execute_commands` into their own methods or stick with the local client/server approach that is currently being used.
> 
@zancas IDK either. Sometimes this seems handy, because you can include a new command easily in both repos, but maybe it's a good idea to extract some of these commands on their own methods...
> I think the succinct API that the current approach has is convenient in terms of boilerplate code that is needed to expose changes in the rust library. The downside is that the app relies on flying strings. like `execute_command(cmd: "sync", "")` and this returns some other string
> 
The pattern for errors is not the best, we need to change it in some point... Checking the start of the string looking for `error` is really rudimentary... I'll happy to change this, but I need this change first from zingolib.
> A more robust API would having a `rustlib.sync()` that returns a `ReturnType` result instead of a string. The downside is that it envolves coding the intermediate code. Maybe the scaffolding created from the udl file can take care of most of it.
> 
Agreed 100%. I added types of all of this returns from the `command` but only when you parse the string to a json object.
> This can be also remediated by creating a better TypeScript API for interacting with `execute_command` and keep this succinct UniFFI API as is to keep a better velocity and less boiler plate for exposing new functionality on the app.
> 
> The main problem is that the Strings resulting on the responses could be broken from a change in the zingolib without that causing a problem in the build itself.
>
In fact, if you review the code base entirely, you can see sometimes, exceptions about this behavior that make you feel even worse...

File: app/LoadedApp/LoadedApp.tsx
Comment (juanky201271) [reply app/LoadedApp/LoadedApp.tsx:1]: Yes, you are right. In the beginning this seemed really straightforward, but after that, with the `unlock` new feature and the different behavior in Android & IOS... it become in `something`, I'll take a look to extract this code like you are saying. But that is not the goal for this PR. Thanks though.

File: ios/RPCModule.swift (hunk 18 above)
Comment (juanky201271) [reply ios/RPCModule.swift]: done.

File: ios/RPCModule.swift (hunk 19 above)
Comment (juanky201271) [reply ios/RPCModule.swift]: I'm using GCD only here for execute commands... and I don't see any problem or collision with RN.

File: ios/RPCModuleBridge.m (hunk 20 above)
Comment (juanky201271) [reply ios/RPCModuleBridge.m]: I changed this, yes, so weird.

File: rust/ios/src/lib.rs (hunk 7 above)
Comment (juanky201271) [reply rust/ios/src/lib.rs:5]: this is not critical.

File: rust/Cargo.toml (hunk 3 above)
Comment (juanky201271) [reply rust/Cargo.toml:12]: I can change this anytime...

File: rust/android/Cargo.toml (hunk 4 above)
Comment (juanky201271) [reply rust/android/Cargo.toml:17]: I can change this anytime...

**Comment by juanky201271** [comment]
@pacu Thanks so much for your help, I did most of your suggestions, and I responded all of your comments. I would appreciate if you can take a look again, just in case. Thanks in advance.

File: ios/RPCModule.swift (hunk 15 above)
Comment (juanky201271) [reply ios/RPCModule.swift]: I changed them, that was weird, thanks.

File: ios/RPCModule.swift (hunk 16 above)
Comment (juanky201271) [reply ios/RPCModule.swift]: That was a mistake traducing the code to swift, I fixed it. Thanks, nice catch.

File: rust/Cargo.toml (hunk 3 above)
Comment (Oscar-Pepper) [reply rust/Cargo.toml:12]: i agree, also "version" isnt even needed here
```suggestion
uniffi = "0.27"
```

File: rust/Cargo.toml [hunk 22]
```diff
@@ -2,15 +2,15 @@
 members = [
     "lib",
     "android",
-    "ios",
+#    "ios",
```
Comment (Oscar-Pepper) [inline rust/Cargo.toml]: this can now be removed

File: rust/Dockerfile [hunk 23]
```diff
@@ -1,13 +1,15 @@
 FROM zingodevops/android_builder:008 as build_android
 
-WORKDIR /opt/zingo/rust/android/
+WORKDIR /opt/zingo/rust/lib/
+# WORKDIR /opt/zingo/rust/android/
```
Comment (Oscar-Pepper) [inline rust/Dockerfile]: remove

File: rust/Dockerfile [hunk 24]
```diff
@@ -1,13 +1,15 @@
...
+
 
 # add the local rust code into the container
 COPY android/ /opt/zingo/rust/android/
 COPY lib/ /opt/zingo/rust/lib/
 COPY Cargo.lock /opt/zingo/rust/Cargo.lock
 COPY Cargo.toml /opt/zingo/rust/Cargo.toml
-COPY ios/ /opt/zingo/rust/ios/
+#COPY ios/ /opt/zingo/rust/ios/
```
Comment (Oscar-Pepper) [inline rust/Dockerfile]: remove

File: rust/android/Cargo.toml (hunk 4 above)
Comment (Oscar-Pepper) [reply rust/android/Cargo.toml:17]: agreed
```suggestion
tokio =  { version = "1.24", features = ["full"] }
```

File: rust/lib/Cargo.toml [hunk 25]
```diff
@@ -10,6 +10,22 @@ zingoconfig = { workspace = true }
 http = "0.2.4"
 lazy_static = "1.4.0"
 base64 = "0.11.0"
+android_logger = "0.11"
+log = "0.4.8"
+uniffi = { workspace = true, features = ["cli"] }
```
Comment (Oscar-Pepper) [inline rust/lib/Cargo.toml]: minor formatting
```suggestion
uniffi = { workspace = true, features = [ "cli" ] }
```

File: rust/lib/Cargo.toml [hunk 26]
```diff
@@ -10,6 +10,22 @@ zingoconfig = { workspace = true }
...
+
+[build-dependencies]
+uniffi_build = { version = "0.27.0", features = [ "builtin-bindgen" ] }
```
Comment (Oscar-Pepper) [inline rust/lib/Cargo.toml]:
```suggestion
uniffi_build = { version = "0.27", features = [ "builtin-bindgen" ] }
```

File: rust/lib/Cargo.toml [hunk 27]
```diff
@@ -10,6 +10,22 @@ zingoconfig = { workspace = true }
...
 
 [dev-dependencies]
 tokio = "1.32.0"
+
+[lib]
+crate-type = ["cdylib", "staticlib"]
+name = "rustlib"
```
Comment (Oscar-Pepper) [inline rust/lib/Cargo.toml:25]: I wonder if a more appropriate name is "zingo" here. I have noticed a `librustlib.so` which doesnt seem right.
"libzingo" and "uniffi_build::generate_scaffolding("src/zingolib.udl") seems better.

BUT I think this is probably best in a following PR. not a requested change!

File: rust/lib/Cargo.toml [hunk 28]
```diff
@@ -10,6 +10,22 @@ zingoconfig = { workspace = true }
 http = "0.2.4"
 lazy_static = "1.4.0"
 base64 = "0.11.0"
+android_logger = "0.11"
+log = "0.4.8"
```
Comment (Oscar-Pepper) [inline rust/lib/Cargo.toml]:
```suggestion
log = "0.4"
```

File: android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt [hunk 29]
```diff
@@ -178,22 +178,24 @@ class ExecuteVersionFromSeed {
...
         assertThat(initFromSeed.seed).isEqualTo(Seeds.ABANDON)
         assertThat(initFromSeed.birthday).isEqualTo(1)
 
-        var version = RustFFI.execute("version", "")
-        System.out.println("\nVersion:")
-        System.out.println(version)
+        val version = uniffi.rustlib.executeCommand("version", "")
+        println("\nVersion:")
+        println(version)
         // we used for zingolib version: `mob-release` & `mob-prerelease`
```
Comment (Oscar-Pepper) [inline android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt]: is this comment still relevant?

File: rust/lib/src/lib.rs [hunk 30]
```diff
@@ -54,6 +62,20 @@ fn construct_uri_load_config(
     Ok((config, lightwalletd_uri))
 }
 
+pub fn init_logging() -> String {
+    // this is only for Android
+    #[cfg(target_os = "android")]
```
Comment (Oscar-Pepper) [inline rust/lib/src/lib.rs:67]: is there a ios version of this function?

File: rust/lib/src/lib.rs [hunk 31]
```diff
@@ -224,7 +246,6 @@ mod tests {
     #[ignore]
     #[tokio::test]
     async fn unconnected_client_framework() {
-        // Use test in RustFFITest.kt as template
```
Comment (Oscar-Pepper) [inline rust/lib/src/lib.rs:227]: this test is failing and not necessary as its covered by a test in offline testsuite
please remove this whole test mod
https://github.com/zingolabs/zingo-mobile/pull/630/files#diff-4fe8b5135dd27e9e8239c61835dafa0713eea327d1f02b33f36d1ed9b17bb129R242-R268

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
This all looks great. I left some minor change requests

File: rust/lib/Cargo.toml (hunk 27 above)
Comment (juanky201271) [reply rust/lib/Cargo.toml:25]: I agree, would be fantastic if you can create an issue with this, thanks in advance.

File: rust/lib/src/lib.rs (hunk 30 above)
Comment (juanky201271) [reply rust/lib/src/lib.rs:67]: nope! in IOS the native logging is working without any action.

File: android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt (hunk 29 above)
Comment (juanky201271) [reply android/app/src/androidTest/java/org/ZingoLabs/Zingo/RustFFITest.kt]: thanks, of course not.

File: rust/lib/src/lib.rs (hunk 31 above)
Comment (juanky201271) [reply rust/lib/src/lib.rs:227]: done! thanks.

File: rust/lib/src/rustlib.udl (hunk 21 above)
Comment (juanky201271) [reply rust/lib/src/rustlib.udl:3]: We need to work in all of these thing, but I don't thing any of them are a blocker. Thanks @pacu

File: ios/AppDelegate.swift (hunk 14 above)
Comment (juanky201271) [reply ios/AppDelegate.swift]: fixed them.

**Review by Oscar-Pepper — APPROVED** [review]
really great work!
